### PR TITLE
Add L2 llama test

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?

L2 llama pretraining CI test

**Collection**: [Note which collection this PR will affect]

# Changelog 
- use pretraining recipe
- apply post-training assertions

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: .github/workflows/cicd-main.yml [hunk 1]
```diff
@@ -3887,6 +3887,38 @@ jobs:
...
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/llama3_pretraining.py \
+        --seq-length 1024 \
+        --devices=2 \
+        --max-steps=6 \
+        --early-stop=3 \
+        --experiment-dir=tests/collections/llm/llama_pretrain_results \
```
Comment (pablo-garay) [inline .github/workflows/cicd-main.yml:3903]: we started - by convention - using /tmp as output directory
AFAIU this is output of the test, so could you update to use /tmp instead

File: .github/workflows/cicd-main.yml [hunk 2]
```diff
@@ -3887,6 +3887,38 @@ jobs:
...
+        --data-path=/home/TestData/nlp/megatron_llama/data/rp2_sample_sentencepiece_preproc_text_document \
+        --tokenizer-path=/home/TestData/nlp/megatron_llama/tokenizer.model \
+        --index-mapping-dir=tests/collections/llm/llama_index_mappings \
+
+        python tests/collections/llm/llama3_pretraining.py \
+        --seq-length 1024 \
+        --devices=2 \
+        --max-steps=6 \
+        --experiment-dir=tests/collections/llm/llama_pretrain_results \
```
Comment (pablo-garay) [inline .github/workflows/cicd-main.yml:3912]: we started - by convention - using /tmp as output directory
AFAIU this is output of the test, so could you update to use /tmp instead

**Comment by maanug-nv** [comment]
might be able to add back check for expected # of ckpts and steps, when _MegatronBatchProgress is fixed.

**Review by ShriyaPalsamudram — DISMISSED** [review]

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 3]
```diff
@@ -300,6 +300,14 @@ def setup(self, trainer: pl.Trainer) -> None:
         self.accelerator.setup(trainer)
         self.trainer = trainer
 
+        try:
+            self.model.optim.lr_scheduler.max_steps = trainer.max_steps
```
Comment (hemildesai) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:304]: Is this the right place to set this? Or does it need to happen before calling `trainer.fit`?

File: tests/collections/llm/llama3_pretraining.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    run.run(pretrain_recipe, direct=True)
+
+    verify_ckpt_dir(
+        pretrain_recipe.log.ckpt,
+        args.early_stop or args.max_steps,
+        pretrain_recipe.trainer.val_check_interval,
+        os.path.join(args.experiment_dir, exp_name),
+    )
+
```
Comment (hemildesai) [inline tests/collections/llm/llama3_pretraining.py:127]: Is it possible to call the `generate` function here to test inference on the generated checkpoint as well?

**Review by hemildesai — COMMENTED** [review]
Looks great, just a couple minor comments.

File: tests/collections/llm/llama3_pretraining.py (hunk 4 above)
Comment (maanug-nv) [reply tests/collections/llm/llama3_pretraining.py:127]: this is a great idea. i can work on adding this in a separate PR.

Comment (maanug-nv) [reply tests/collections/llm/llama3_pretraining.py:127]: i don't know if the checkpoint would need any conversion, do you know?

File: nemo/lightning/pytorch/strategies/megatron_strategy.py (hunk 3 above)
Comment (maanug-nv) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:304]: Looks like this [line](https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/pytorch/optim/base.py#L88) is triggered by `setup_megatron_parallel`->`setup_optimizers`, [which](https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/pytorch/strategies/megatron_strategy.py#L313) starts later in Strategy `setup()`. I checked with breakpoint to make sure scheduler's max_steps is correct when the scheduler setup runs. So I think this is early enough. I'll add an assertion at train start for this

Comment (maanug-nv) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:304]: see 58c44363997f06544109ad78333c2b8825225cb7

**Review by hemildesai — APPROVED** [review]

**Review by hemildesai — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]
